lib/kube/proxy/portforward_websocket.go

```diff
@@ -0,0 +1,339 @@
+		}
+		for _, p := range strings.Split(portString, ",") {
+			port, err := parsePortString(p)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			ports = append(ports, port)
+		}
+	}
+
+	return ports, nil
+}
+
+// websocketChannelPair represents the error and data streams for a single
+// port.
+type websocketChannelPair struct {
+	port        uint16
+	podName     string
+	requestID   string
+	dataStream  io.ReadWriteCloser
+	errorStream io.ReadWriteCloser
+}
+
+func (w *websocketChannelPair) close() {
+	w.dataStream.Close()
+	w.errorStream.Close()
+}
+
+func (w *websocketChannelPair) sendErr(err error) {
+	if err == nil {
+		return
+	}
+	fmt.Fprintf(w.errorStream, "error forwarding port %d to pod %s: %v", w.port, w.podName, err)
+}
+
+// websocketPortforwardHandler is capable of processing a single port forward
+// request over a websocket connection
+type websocketPortforwardHandler struct {
+	conn          *wsstream.Conn
+	streamPairs   []*websocketChannelPair
+	podName       string
+	targetConn    httpstream.Connection
+	onPortForward portForwardCallback
+	logrus.FieldLogger
+	context context.Context
+}
+
+// run invokes the targetConn SPDY connection and copies the client data into
+// the targetConn and the responses into the targetConn data stream.
+// If any error occurs, stream is closed an the error is sent via errorStream.
+func (h *websocketPortforwardHandler) run() {
+	wg := sync.WaitGroup{}
+	wg.Add(len(h.streamPairs))
+
+	for _, pair := range h.streamPairs {
+		p := pair
+		go func() {
+			defer wg.Done()
+			h.portForward(p)
+		}()
+	}
+
+	wg.Wait()
+}
+
+// portForward copies the client and upstream streams.
+func (h *websocketPortforwardHandler) portForward(p *websocketChannelPair) {
+	h.Debugf("Forwarding port %v -> %v.", p.requestID, p.port)
+	h.forwardStreamPair(p)
+
+	h.Debugf("Completed forwarding port %v -> %v.", p.requestID, p.port)
+}
+
+func (h *websocketPortforwardHandler) forwardStreamPair(p *websocketChannelPair) {
+	// create error stream
+	headers := http.Header{}
+	headers.Set(StreamType, StreamTypeError)
+	headers.Set(PortHeader, fmt.Sprint(p.port))
+	headers.Set(PortForwardRequestIDHeader, p.requestID)
+
+	// read and write from the error stream
+	targetErrorStream, err := h.targetConn.CreateStream(headers)
+	h.onPortForward(fmt.Sprintf("%v:%v", h.podName, p.port), err == nil /* success */)
+	if err != nil {
+		p.sendErr(err)
+		p.close()
+		return
+	}
+	defer targetErrorStream.Close()
+	wg := &sync.WaitGroup{}
+	wg.Add(2)
+
+	go func() {
+		defer wg.Done()
+		_, err := io.Copy(targetErrorStream, p.errorStream)
+		if err != nil && !errors.Is(err, io.EOF) {
+			h.Debugf("Copy stream error: %v.", err)
+		}
+	}()
+
+	errClose := make(chan struct{})
+	go func() {
+		defer wg.Done()
+		defer close(errClose)
+		_, err := io.Copy(p.errorStream, targetErrorStream)
+		if err != nil && !errors.Is(err, io.EOF) {
+			h.Debugf("Copy stream error: %v.", err)
+		}
+	}()
```

Comment:
Can we use [lib/utils.ProxyConn](https://github.com/gravitational/teleport/blob/master/lib/utils/proxyconn.go#L31) here instead of reimplementing the double-copy-loop logic?

Below too.

Reply:
Nice. Didn't know that